Vocabulary sync (#694)
blindnet-io/product-management: refs/high-level-architecture/readme.md

| :------------ | :---------------------------------------------------------------- |
| **PR #**      | [596](https://github.com/blindnet-io/product-management/pull/596), [694](https://github.com/blindnet-io/product-management/pull/694) |
| **Author(s)** | milstan (milstan@blindnet.io)                                     |
| **Updated**   | 2022-06-14                                                        |

## Objective

This document is about:

- the context in which [blindnet devkit](../lexicon/RFC-Lexicon-2.md#blindnet-devkit) has to operate, and
- the structure of components of [blindnet devkit](../lexicon/RFC-Lexicon-2.md#blindnet-devkit) (in relation to that context).

This document complements the [High Level Conceptualization](../high-level-conceptualization/).
Its goal is to propose the optimal component structure of the product.

## Motivation

The [High Level Conceptualization](../high-level-conceptualization/) is insufficient for understanding of the context of use. We need more clarity about the data capture lifecycle and the parties involved.

Also, having the right structure of components is very important for several reasons, including for:

> - client [Systems](../lexicon/RFC-Lexicon-2.md#system) to be able to use only a part of our system,

Reviewer: ```suggestion
- client [Systems](../lexicon/RFC-Lexicon-2.md#system) to be able to use only a part of [blindnet devkit](../lexicon/RFC-Lexicon-2.md#blindnet-devkit),
```

> - our system to be interoperable with 3rd party solutions (that cover a sub-scope of our scope or that cover requirements adjacent to our scope),

Reviewer: ```suggestion
- [blindnet devkit](../lexicon/RFC-Lexicon-2.md#blindnet-devkit) to be interoperable with 3rd party solutions (that cover a sub-scope of our scope or that cover requirements adjacent to our scope),
```

## Proposal - Components

All components are presented in their roles in the following, more detailed image:

<img width="1400" alt="HIGH-LEVEL ARCHITECTURE DETAIL" src="./img/HLA-detail.png">

### Capture Component

The Capture Component is the entry point of the product. It is responsible for putting in motion [Data Captures](../high-level-conceptualization/README.md#the-main-concept---data-capture) in a way that reinforces privacy by ensuring confidentiality and control.

> It is responsible for encrypting the data, obtaining [Consent](../high-level-conceptualization/README.md#data-capture--consent), and generating Data Capture metadata.

Author: @m4rk055 See the new wording.

This component also provides functions allowing to support Data Consumers in situations of lost access.

The component MAY accomplish those responsibilities through coordination with user identity management and encryption tools (external or not).

### Privacy Computation Engine

Privacy Computation Engine is a component that client [Systems](../lexicon/RFC-Lexicon-2.md#system) CAN run in order to interpret their rights to hold and treat a particular [Data Capture](../high-level-conceptualization/README.md#the-main-concept---data-capture) at a particular point of time.
The engine CAN also calculate a response to [Data Subjects'](../lexicon/RFC-Lexicon-2.md#data-subject) [Privacy Requests](../lexicon/RFC-Lexicon-2.md#privacy-request).

A Privacy Computation Engine component is composed of three elements:

- a [Privacy Compiler](#privacy-compiler)
- a [Privacy Request Capture Interface](#privacy-request-capture-interface)
- a [Customization API](#customization-api)

#### Privacy Compiler

> A **_Privacy Compiler_** computes one or more actions (and to them association motivations/reasons/explanations) that need to be performed by a [Systems](../lexicon/RFC-Lexicon-2.md#system) in order to respect [Data Subjects'](../lexicon/RFC-Lexicon-2.md#data-subject) data rights.

Reviewer 2: ```suggestion
A **_Privacy Compiler_** computes one or more actions (and to them association motivations/reasons/explanations) that need to be performed by a [System](../lexicon/RFC-Lexicon-2.md#system) in order to respect [Data Subjects'](../lexicon/RFC-Lexicon-2.md#data-subject) data rights.
```

Appropriate abstractions SHOULD be designed on top of key components for (at least, not limited to) user identity, access management, encryption, Data Rights Management, storage, and Privacy Request capture.
These abstractions SHOULD be inspired by the [Bridge Pattern](https://en.wikipedia.org/wiki/Bridge_pattern).

This interoperability SHOULD support the following (extreme) example scenarios:

- A System that uses Keycloak for user identity management, CAN use our Data Capture component in combination with 3rd party Encryption library. The System CAN instantiate and use our Privacy Computation Engine, and implement their own Storage able to automatically interpret and enforce actions resulting as output of the DPrivacy Computation Engine (automatic deletion and modification).

- A System that is essentially a wordpress website, CAN embed the Capture Component in a wordpress webform, use the Storage Component, the Data Consumer Interface (as separate from the wordpress website), use Privacy Computation Engine and embed the Privacy Request Capture Interface in the wordpress website respecting its look and feel.

> - A System that has their own user identity management, uses 3rd party e2ee library, has its own storage and interfaces for data view, CAN instantiate the Privacy Computation Engine and use only the parts of the Data Consumer Interface related to Privacy Request parametrization and approval.

Reviewer 2: ```suggestion
- A System that has its own user identity management, uses 3rd party e2ee library, has its own storage and interfaces for data view, CAN instantiate the Privacy Computation Engine and use only the parts of the Data Consumer Interface related to Privacy Request parametrization and approval.
```

### Separation of Form and Function

> The components of [blindnet devkit](../lexicon/RFC-Lexicon-2.md#blindnet-devkit) MUST be made in such a way that a client [Systems](../lexicon/RFC-Lexicon-2.md#system) CAN use only the functional part and implement its own interfaces or otherwise achieve its own desired look and feel and UX.

Reviewer 2: ```suggestion
The components of [blindnet devkit](../lexicon/RFC-Lexicon-2.md#blindnet-devkit) MUST be made in such a way that client [Systems](../lexicon/RFC-Lexicon-2.md#system) CAN use only the functional part and implement its own interfaces or otherwise achieve its own desired look and feel and UX.
```
